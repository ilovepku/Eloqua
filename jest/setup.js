import 'react-native-gesture-handler/jestSetup';
import mockAsyncStorage from '@react-native-async-storage/async-storage/jest/async-storage-mock';

jest.mock('react-native-reanimated', () => {
  const Reanimated = require('react-native-reanimated/mock');

  // The mock for `call` immediately calls the callback which is incorrect
  // So we override it with a no-op
  Reanimated.default.call = () => {};

  return Reanimated;
});

jest.mock('react-native-track-player', () => {
  return {
    addEventListener: jest.fn(),
    registerEventHandler: jest.fn(),
    registerPlaybackService: jest.fn(),
    setupPlayer: jest.fn(),
    destroy: jest.fn(),
    updateOptions: jest.fn(),
    add: jest.fn(),
    remove: jest.fn(),
    skip: jest.fn(),
    skipToNext: jest.fn(),
    skipToPrevious: jest.fn(),
    removeUpcomingTracks: jest.fn(),
    // playback commands
    reset: jest.fn(),
    play: jest.fn(),
    pause: jest.fn(),
    stop: jest.fn(),
    seekTo: jest.fn(),
    setVolume: jest.fn(),
    setRate: jest.fn(),
    // player getters
    getQueue: jest.fn(),
    getTrack: jest.fn(),
    getCurrentTrack: jest.fn(),
    getVolume: jest.fn(),
    getDuration: jest.fn(),
    getPosition: jest.fn(),
    getBufferedPosition: jest.fn(),
    getState: jest.fn(),
    getRate: jest.fn(),
  };
});

jest.mock('@react-native-async-storage/async-storage', () => mockAsyncStorage);

jest.mock('react-native-snackbar', () => {
  return {
    show: jest.fn(),
  };
});

// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
