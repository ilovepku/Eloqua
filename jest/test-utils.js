import React from 'react'
import {createStore} from 'redux'
import {Provider} from 'react-redux'
import {act, render as rtlRender} from '@testing-library/react-native'

import reducer from '../src/redux/rootReducer'

// wait for response helper
export async function wait(ms = 0) {
  await act(
    () =>
      new Promise(resolve => {
        setTimeout(resolve, ms)
      }),
  )
}

// override default render with mock store
function render(
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {},
) {
  function Wrapper({children}) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, {wrapper: Wrapper, ...renderOptions})
}

// re-export everything
export * from '@testing-library/react-native'
// override render method
export {render}
