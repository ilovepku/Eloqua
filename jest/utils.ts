import {act} from '@testing-library/react-native';

// wait for response helper
export async function wait(ms = 0) {
  await act(() => {
    return new Promise((resolve) => {
      setTimeout(resolve, ms);
    });
  });
}
