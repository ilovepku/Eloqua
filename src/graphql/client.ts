import { ApolloClient, InMemoryCache } from '@apollo/client';

const client = new ApolloClient({
    uri: 'https://frozen-river-77426.herokuapp.com/query',
    cache: new InMemoryCache()
});

export default client