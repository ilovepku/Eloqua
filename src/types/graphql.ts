// ====================================================
// GraphQL query operation: AllPiecesQuery
// ====================================================

export interface AllPiecesQuery_piece {
  __typename?: 'pieces'
  id: number
  name: string
  person_id: number
  person: {
    __typename?: 'persons'
    name: string
    img_filename: string
  }
  date: string
  duration: number
  audio_filename: string
  piece_categories: {
    category_id: number
  }[]
  text: string
}

export interface AllPiecesQuery {
  search: AllPiecesQuery_piece[]
}

// ====================================================
// GraphQL query operation: CategoriesQuery
// ====================================================

export interface CategoriesQuery_category {
  __typename?: 'categories'
  id: number
  name: string
  icon_filename: string
}

export interface CategoriesQuery {
  search: CategoriesQuery_category[]
}

// ====================================================
// GraphQL query operation: PersonsQuery
// ====================================================

export interface PersonsQuery_person {
  __typename?: 'persons'
  id: number
  name: string
  img_filename: string
}

export interface PersonsQuery {
  search: PersonsQuery_person[]
}
