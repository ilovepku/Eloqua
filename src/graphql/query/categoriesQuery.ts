import {gql} from '@apollo/client'

const categoriesQuery = gql`
  query categoriesQuery {
    categories(limit: 11) {
      id
      name
      icon_filename
    }
  }
`

export default categoriesQuery
