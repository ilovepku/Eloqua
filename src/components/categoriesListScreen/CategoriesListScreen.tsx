import React from 'react';
import {View, ScrollView} from 'react-native';
import {useQuery} from '@apollo/client';
import tailwind from 'tailwind-rn';

import categoriesQuery from '../../graphql/query/categoriesQuery';
import {CategoriesQuery_category} from '../../types/graphql';

import Error from '../error/Error';
import Loading from '../loading/Loading';
import CategoryItem from './CategoryItem';

export default function CategoriesListScreen() {
  const {loading, error, data} = useQuery(categoriesQuery);

  if (loading) {
    return <Loading />;
  }
  if (error) {
    return <Error errMsg={error.message} />;
  }

  return (
    <View style={tailwind('flex-1 bg-white')}>
      <ScrollView
        contentContainerStyle={tailwind(
          'flex-row flex-wrap justify-evenly pb-4',
        )}>
        {data.categories.map((category: CategoriesQuery_category) => (
          <CategoryItem key={`category-${category.id}`} category={category} />
        ))}
      </ScrollView>
    </View>
  );
}
