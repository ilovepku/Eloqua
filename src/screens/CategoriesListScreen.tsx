import React from 'react'
import {View, ScrollView} from 'react-native'
import {useQuery} from '@apollo/client'
import tailwind from 'tailwind-rn'

import categoriesQuery from '../graphql/query/categoriesQuery'
import {CategoriesQueryCategory} from '../types/graphql'

import Error from '../components/Error'
import Loading from '../components/Loading'
import CategoryItem from '../components/CategoryItem'

const CategoriesListScreen: React.FC = () => {
  const {loading, error, data} = useQuery(categoriesQuery)

  if (loading) {
    return <Loading />
  }
  if (error) {
    return <Error errMsg={error.message} />
  }

  return (
    <View style={tailwind('flex-1 bg-white')}>
      <ScrollView
        contentContainerStyle={tailwind(
          'flex-row flex-wrap justify-evenly pb-4',
        )}
      >
        {data.categories.map((category: CategoriesQueryCategory) => (
          <CategoryItem key={`category-${category.id}`} category={category} />
        ))}
      </ScrollView>
    </View>
  )
}

export default CategoriesListScreen
