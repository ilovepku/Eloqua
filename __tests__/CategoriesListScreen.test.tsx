import React from 'react'
import {render} from '@testing-library/react-native'
import {MockedProvider} from '@apollo/client/testing'

import {wait} from '../jest/test-utils'
import {CategoriesQueryCategory} from '../src/types/graphql'
import categoriesQuery from '../src/graphql/query/categoriesQuery'
import CategoriesListScreen from '../src/screens/CategoriesListScreen'

const mockLoadingMsg = 'Loading...'
const mockErrMsg = 'mock error'

const categoriesQueryMock = {
  request: {
    query: categoriesQuery,
  },
  result: {
    data: {
      categories: [
        {
          id: 1,
          name: 'Acknowledgement',
          icon_filename: 'acknowledgement.png',
        },
        {
          id: 2,
          name: 'Commencement',
          icon_filename: 'commencement.png',
        },
      ],
    },
  },
}
const categoriesQueryErrorMock = {
  request: {
    query: categoriesQuery,
  },
  error: new Error(mockErrMsg),
}

jest.mock('../src/components/loading/Loading', () => () => (
  <p>{mockLoadingMsg}</p>
))
jest.mock(
  '../src/components/error/Error',
  () => ({errMsg}: {errMsg: string}) => <p>{errMsg}</p>,
)
jest.mock(
  '../src/components/categoriesListScreen/CategoryItem',
  () => ({
    category: {name, id, icon_filename},
  }: {
    category: CategoriesQueryCategory
  }) => (
    <>
      <p>{name}</p>
      <p>{id}</p>
      <p>{icon_filename}</p>
    </>
  ),
)

describe('CategoriesListScreen', () => {
  it('should render loading state initially', () => {
    const tree = render(
      <MockedProvider mocks={[categoriesQueryMock]} addTypename={false}>
        <CategoriesListScreen />
      </MockedProvider>,
    ).toJSON()
    expect(tree?.children).toContain(mockLoadingMsg)
  })

  it('should show error UI', async () => {
    const component = render(
      <MockedProvider mocks={[categoriesQueryErrorMock]} addTypename={false}>
        <CategoriesListScreen />
      </MockedProvider>,
    )

    await wait()

    const tree = component.toJSON()
    expect(tree?.children).toContain(mockErrMsg)
  })

  it('renders correctly', async () => {
    const component = render(
      <MockedProvider mocks={[categoriesQueryMock]} addTypename={false}>
        <CategoriesListScreen />
      </MockedProvider>,
    )

    await wait()

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
