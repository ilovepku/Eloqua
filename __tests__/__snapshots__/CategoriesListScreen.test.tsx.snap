// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesListScreen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 1)",
      "flexBasis": "0%",
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "space-evenly",
        "paddingBottom": 16,
      }
    }
  >
    <View>
      <p>
        Acknowledgement
      </p>
      <p>
        1
      </p>
      <p>
        acknowledgement.png
      </p>
      <p>
        Commencement
      </p>
      <p>
        2
      </p>
      <p>
        commencement.png
      </p>
    </View>
  </RCTScrollView>
</View>
`;
